### ISSUE TITLE
refit method throws exception

### ISSUE THREAD

**Issue by lpatruno** [automl/auto-sklearn#440]
Hi folks,

I'm trying to test out auto-sklearn and was able to fit some random forest models with cv. When I refit the model over the entire dataset though, I get the following error:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/opt/conda/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

/opt/conda/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    393                             if callable(meth):
    394                                 return meth(obj, self, cycle)
--> 395             return _default_pprint(obj, self, cycle)
    396         finally:
    397             self.end_group()

/opt/conda/lib/python3.6/site-packages/IPython/lib/pretty.py in _default_pprint(obj, p, cycle)
    508     if _safe_getattr(klass, '__repr__', None) is not object.__repr__:
    509         # A user-provided repr. Find newlines and replace them with p.break_()
--> 510         _repr_pprint(obj, p, cycle)
    511         return
    512     p.begin_group(1, '<')

/opt/conda/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    699     """A pprint that just redirects to the normal repr function."""
    700     # Find newlines and replace them with p.break_()
--> 701     output = repr(obj)
    702     for idx,output_line in enumerate(output.splitlines()):
    703         if idx:

/opt/conda/lib/python3.6/site-packages/sklearn/base.py in __repr__(self)
    286     def __repr__(self):
    287         class_name = self.__class__.__name__
--> 288         return '%s(%s)' % (class_name, _pprint(self.get_params(deep=False),
    289                                                offset=len(class_name),),)
    290 

/opt/conda/lib/python3.6/site-packages/sklearn/base.py in get_params(self, deep)
    226         """
    227         out = dict()
--> 228         for key in self._get_param_names():
    229             # We need deprecation warnings to always be on in order to
    230             # catch deprecated param values.

/opt/conda/lib/python3.6/site-packages/sklearn/base.py in _get_param_names(cls)
    207                                    " %s with constructor %s doesn't "
    208                                    " follow this convention."
--> 209                                    % (cls, init_signature))
    210         # Extract and sort argument names excluding 'self'
    211         return sorted([p.name for p in parameters])

RuntimeError: scikit-learn estimators should always specify their parameters in the signature of their __init__ (no varargs). <class 'autosklearn.automl.AutoMLClassifier'> with constructor (self, *args, **kwargs) doesn't  follow this convention.
```

Any ideas?

**Comment by mfeurer** [comment]
Do you have a minimal working example that works together with one of the scikit-learn example datasets?

**Comment by lpatruno** [comment]
Hi @mfeurer . I'm not sure what you mean by your question. Can you rephrase?

**Comment by mfeurer** [comment]
Sorry for the late answer, I'm not receiving any emails for this repo any more and didn't notice this.

In order to find out what's going wrong here I would like to reproduce the error on my local machine. Therefore, it would be good if you could provide me with a minimal example which triggers this error. In the optimal case, this is with a datasets that I can programatically retrieve, for example one of the [scikit-learn example datasets](http://scikit-learn.org/stable/datasets/index.html#toy-datasets).

**Comment by lpatruno** [comment]
Hi @mfeurer ,

Thanks for your response. Question, are you no longer actively maintaining this repository?

I do not have a toy example (I was training models on proprietary datasets), but I can provide some more context on the issue. I'm not sure if this helps, but it's about as much as I can provide at the moment.

The following code ran without any exceptions

```
X_train = ...
y_train = ...

automl = autosklearn.classification.AutoSklearnClassifier(
    time_left_for_this_task=120,
    per_run_time_limit=30,
    tmp_folder='/tmp/autosklearn_cv_example_tmp',
    output_folder='/tmp/autosklearn_cv_example_out',
    delete_tmp_folder_after_terminate=False,
    resampling_strategy='cv',
    resampling_strategy_arguments={'folds': 3},
    include_estimators=["random_forest"]
)

# Create feature type list from openml.org indicator and run autosklearn
feat_type = ['Categorical' if col in categorical_feats else 'Numerical' for col in X_cols]

automl.fit(X_train.copy(), y_train.copy(), feat_type=feat_type)

print(automl.show_models())

print(automl.sprint_statistics())
```

However, the following line through the exception from my first comment:

```
automl.refit(X_train.copy(), y_train.copy())
```

Again, apologies if this is not enough context. Let me know if there is anything else I can provide.

**Comment by mfeurer** [comment]
I'm from time to time spending some time on Auto-sklearn ;) If there's a PR, I'll try to merge it as fast as possible. Unfortunately, I don't have the time myself to follow up on all the bug reports, but @ahn1340 will help out in the future, so the situation should improve.

We'll have a look whether we can reproduce the issue, it might take a bit of time until we get to this, though.

**Comment by lpatruno** [comment]
Understood @mfeurer . Thanks for having a look.

If I get some more time on my end, I'll work on trying to reproduce with an sklearn dataset.

**Comment by rquintino** [comment]
same error on refit here, running https://github.com/automl/auto-sklearn/blob/master/example/example_crossvalidation.py from docker image (mfeurer/auto-sklearn:latest)
Python 3.5.2 (default, Nov 23 2017, 16:37:01)

**Comment by ahn1340** [comment]
Hi @lpatruno,

@mfeurer and I could reproduce the error message that you posted. The problem was that Ipython automatically prints the return value when a method or a function is called without assigning its return value to a variable. Hence, the error message should disappear either if you run auto-sklearn with normal python (which does not automatically print the return value like Ipython), or if you assign a variable to the function call like so:

`foo = automl.refit(X_train.copy(), y_train.copy())`

There are some other issues regarding the refit method, and we are currently working on them for the next stable release of auto-sklearn, so hopefully this problem will be removed in the next release.  For the time being, my suggestion above should allow you to use auto-sklearn without problem. If you encounter any other issues, please let us know.

**Comment by lpatruno** [comment]
Hi @ahn1340 ,

Thanks for your message. I've rerun my script and it runs without errors. Much appreciated! Thanks @mfeurer as well!

**Comment by mfeurer** [comment]
This is now fixed in the development branch via #476.
